### DISCUSSION TITLE
gonew: templates for new modules

### DISCUSSION THREAD

**Discussion by rsc** [golang/go#61669] [Discussions]
A common request we hear from users is some kind of 'go new' functionality to start a new module with some kind of basic template.

It seems like a tool along these lines should not be limited to templates provided by the Go team: it should be possible for anyone to define templates. That is, templates should just be ordinary modules. At that point the core job of 'go new' is mainly to download a template, change its module path, and drop it in a new directory for editing. There may be more of course, but that's the core.

After literally years of occasionally batting this idea around with @Sajmani, a few months ago I wrote a very trivial text-substitution-based, untested tool at rsc.io/tmp/gonew that did this and mentioned it to a few other people. 

Almost immediately, I found another use: a book author emailed asking how best to make book examples available without requiring book readers to puzzle through the vagaries of cloning a git repo. The obvious answer is gonew, which lets people use only Go tools:

    $ gonew book.com/mybook-examples

(In this form there's no changing the module path, just clone and drop in a new directory for editing.) 

I've also found that form useful for grabbing a copy of some module I am using, to grep through code or to hack up and add to a go.work, like a very lightweight version of github.com/rogpeppe/gohack.

A couple teams at Google were also interested to try gonew for templating. The ServiceWeaver team was interested enough that they put `go install rsc.io/tmp/gonew@latest` in their getting started template docs. That both confirmed interest and scared me enough (tmp is in the name for a reason!) that I wrote a small, real, tested version at golang.org/x/tools/cmd/gonew, which I've just submitted. 

That tool, golang.org/x/tools/cmd/gonew, is experimental in the sense that we aren't sure what will happen next, but it is non-experimental in the sense that we promise to keep the core command line that exists today working. That way, people can confidently post scripts like [ServiceWeaver now does](https://github.com/ServiceWeaver/template):

    $ go install golang.org/x/tools/cmd/gonew@latest
    $ gonew github.com/ServiceWeaver/template example.com/foo

I've also updated the [golang.org/x/example](https://go.googlesource.com/example) repo to work with gonew, so here are three sample examples:

    $ gonew golang.org/x/example/hello
    $ gonew golang.org/x/example/helloserver
    $ gonew golang.org/x/example/outyet

I am confident there is plenty that template authors will want from 'gonew' beyond this basic functionality. I'd like to hear more about that in the comments. We will try to identify general mechanisms that will satisfy many of the things people need. 

Another interesting thing about having these kinds of templates is that, with a clear way to identify them (TBD), you can imagine good support for searching for templates on pkg.go.dev and for instantiating templates in the IDE experience using gopls. If you can think of other ways that the overall Go tooling ecosystem might be changed to incorporate the idea of templates, please post them too.

This is the very _very_ beginning of what could turn out to be an important concept or tool in the Go ecosystem. Perhaps eventually it will graduate to 'go new'. It's a bit unusual for us to publish such a tiny idea, but the core functionality seems useful enough to be worth taking the community's attention to see how it should evolve.

Thanks for any feedback, and please keep the discussion respectful.

**Comment by oherych** [comment]
So I will start with a small story. I had a popular case when we had to split a monolith application into smaller parts. At the same time, several teams should do this in parallel. Most of them had zero experience with Golang. I made the decision to create a small tool for creating a project from a templates.  

Actually, we don't use it as we lost this project :). But I would like to share draft. Yes, code is pretty raw but it can demonstrate some principles.

https://github.com/oherych/project

**The main ideas**:
1) **Template** should be configurable - So, we want to have possible to pass some settings like a project name, database, protocols, etc.
2) **Template** is application - We want to be flexible instead just copy files from template project (ex GithHub templates). Ideas using DSL or configuration based solution was bad.
3) **Template** has main function - as usual application it can be runned. 
4) **Template** has **Jobs**. Each Job can have Up and Down handler. So if Job failed we can rollback all steps.
5) Each **Job** has name and nice description
6) **Jobs** are crossplatform.
7) **Job** is interface and can be implemented by end user

**Examples of Jobs**:
- Create empty file (Touch)
- Create file from String / Bytes / Template 
- Create folder


```go
func prepareJobs(serviceName string) []project.Job {
	const perm = 0777

	b := project.NewBuilder()
	b.DirIn("cmd", perm)
	b.Template("main.go", perm, templateMain, map[string]interface{}{
		"serviceName": serviceName,
		"port":        ":8080",
	})

	return b.Done()
}
```


I hope our experance will be helpful for you.

**Comment by iflamed** [comment]
May should support workflow config by `YAML` or `TOML`, to add  `beforeCreate` and `postCreate` command.

I use `gonew` to create project from template [github.com/caitunai/go-blueprint](https://github.com/caitunai/go-blueprint), after project created, I have to update the `.golangci.yaml` by hand.

**Comment by earthboundkid** [comment]
Another example to look at: https://autostrada.dev/ from @alexedwards.

**Comment by mdhender** [comment]
I set up a structure similar to to golang.org/x/example and created my templates as subdirectories. Each template had a `COPYING` file (not a `LICENSE` file) and the root of my repository had a `LICENSE` file. It surprised me that the `gonew` command copied the `LICENSE` in from the root. Not a big deal (the work around is to remove the `LICENSE` file from the root). Other than that, it's nice to work with.

**Comment by joeshaw** [comment]
Here at work we have a `go-api-template` project that we recommend folks use if they need to build a new service that will expose internal or external APIs.  It sets up some basic handlers for healthchecks, adds in middleware for logging and panic handling and tracing, and defines some types that we commonly use for things like errors.  `gonew` looks like it'd be a good fit for the code side of things.

However, our template also handles a bunch of scaffolding outside of the code itself for our CI system and deployment to Kubernetes.  For this we need templating outside of simply adjusting package paths in Go code and `go.mod`.  We're currently using cruft for this -- I don't love having a Python dependency for our Go projects, especially since we're not otherwise a Python shop -- and it'd be nice if `gonew` handled these cases in a relatively straightforward way.

**Comment by dghubble** [comment]
I think most developers, at some point, write a little tool like this. And are briefly enamored with them. Rather than everyone linking their tool or advocating their templating preferences, it might be better to think about **what `gonew` could bring to the Go language ecosystem** - Plain project templater tools, we have hundreds to choose from. But I can see an argument that Rust and NPM language tools use project templating to elevate use of the language itself.

**Reply by dghubble** [reply]
A sticking point of every templating tool (and why there are so many) is that they boil down to trying to standardize something folks are quite opinionated about. And this is a pretty crowded landscape

**Reply by jimmyfrasche** [reply]
> I can see an argument that Rust and NPM language tools use project templating to elevate use of the language itself.

I'm unfamiliar with them. What do they do differently from other solutions that does this?

**Reply by dghubble** [reply]
Reaching out to those communities would be wise. I can't claim to speak for them, but as a Gopher who casually uses other languages, I can say they've invested substantial effort into integrating those tools into docs and tutorials, making those tools fit major project's use cases,  and winning developer good will. In some cases, (e.g. cargo) aided by being there early on (whereas gonew is quite late).

The main point is, `gonew` should have clarity around how it aims to enhance the Go ecosystem. Onus is on gonew. Otherwise it becomes another templater project.

**Reply by dghubble** [reply]
Some fruitful questions (which I don't know the answers to) might be: 

* How/where are other languages / package systems using builtin template tools to good effect? What have they gained / regretted?
* What can you add as an "official" language Go tool that other tools cannot? Maybe that's just reuse of the VCS discovery / goproxy / checksum machinery. Or something more advanced, idk

**Comment by leo-costa** [comment]
For what it's worth, .NET CLI offers a very good story here. With `dotnet new` you can

- Start a new project using a template name
- Search, install, and update templates from any Nuget source
- List all your templates.
- The template themselves can have their own options or arguments.

Examples:

To start a console project -> `dotnet new console` 
Web API project without OpenAPI support -> `dotnet new webapi --no-openapi`
Start AWS lambda project -> `dotnet new lambda.EmptyFunction` (given you had installed AWS lambda template)


This is the current help page on what you can accomplish with it.

```
Description:
  Template Instantiation Commands for .NET CLI.

Usage:
  dotnet new [<template-short-name> [<template-args>...]] [options]
  dotnet new [command] [options]

Arguments:
  <template-short-name>  A short name of the template to create.
  <template-args>        Template specific options to use.

Options:
  -o, --output <output>    Location to place the generated output.
  -n, --name <name>        The name for the output being created. If no name is specified, the name of the output directory is used.
  --dry-run                Displays a summary of what would happen if the given command line were run if it would result in a template creation.
  --force                  Forces content to be generated even if it would change existing files.
  --no-update-check        Disables checking for the template package updates when instantiating a template.
  --project <project>      The project that should be used for context evaluation.
  -v, --verbosity <LEVEL>  Sets the verbosity level. Allowed values are q[uiet], m[inimal], n[ormal], and diag[nostic]. [default: normal]
  -d, --diagnostics        Enables diagnostic output.
  -?, -h, --help           Show command line help.

Commands:
  create <template-short-name> <template-args>  Instantiates a template with given short name. An alias of 'dotnet new <template name>'.
  install <package>                             Installs a template package.
  uninstall <package>                           Uninstalls a template package.
  update                                        Checks the currently installed template packages for update, and install the updates.
  search <template-name>                        Searches for the templates on NuGet.org.
  list <template-name>                          Lists templates containing the specified template name. If no name is specified, lists all templates.
```

**Comment by dolmen** [comment]
I want Go templates for [go.dev/play](https://go.dev/play).

So far writing code on go.dev/play involving multiple files (full module, with fixed or flexible (`@latest`, `@master`) dependencies) involves starting from scratch from the ["Multiple files"](https://go.dev/play/p/BWJ4dcUqVfT) template.

It would be helpful to have a parameter for [go.dev/play](https://go.dev/play) to specify a module as template.

**Comment by dolmen** [comment]
For my Go projects hosted on GitHub, my current need is management tools to handle CI using Github Actions:
* setup CI (template can help, but I usually I already have my Go code ready when I want to add CI)
* upgrade CI setup
* handle CI for multi-module repos (actions/setup-go doesn't yet have flexible support)

In general I want more tooling for handling multi-modules Git repositories (I have some work in progress in a [Makefile](https://github.com/blueboardio/go-blobqueue/blob/master/Makefile) that I copy from project to project).

**Comment by kamilsk** [comment]
In my case, I use templates to reduce the time for setting up infrastructure issues such as ci/cd, publishing documentation, and installing related tools. I have context-specific templates, for example, simple [go-module](https://github.com/octomation/go-module) or [go-tool](https://github.com/octomation/go-tool) to create a new CLI tool. I tested `go new` with my case, which looks promising. I wrote detailed research here https://gist.github.com/kamilsk/159eef5114f4e67b8a45cbcf76f80798.

I hope it will be as simple as possible and helps the community avoid boring stuff, not complex issues (debugging and maintaining that will be a huge pain). Also, some potential templates can be found here https://awesome-go.com/project-layout/.

**Comment by bkielbasa** [comment]
In my team, we have a simple UI where I can put the name of the project and some additional features that I can choose when starting a new project like:

* add DB support (it adds to the source code everything that's needed to access the DB with basic migrations)
* add redis support
* add message bus (RabbitMQ in my case)

It'd be awesome to be able to create similar functionalities with `gonew`. I can see the usage like this

```sh
gonew github.com/path/template.git github.com/myorg/myproject --param redis=yes --param db=yes
```

or something similar. The idea is simple: we are able to add high-level parameters that help us customize the created project. It should be able to edit(template) files and create/skip specific files based on those parameters.

Those features should be good enough to do what, for example, @kamilsk said to customize any infrastructure-related things like CI/CD, generating docker-compose files, and much more.

```sh
gonew github.com/path/template.git github.com/myorg/myproject --param ci=githubactions
```

**Comment by zhb127** [comment]
gonew can keep template files default permissions?

this is my test template: [https://github.com/zhb127/gonew-test](https://github.com/zhb127/gonew-test)
I have permission to exec my script:

```bash
./helloserver/scripts/init-local-dev.sh

start local db...
start local redis...
The local development environment is initialized.
```

create project from template: 

```bash
gonew github.com/zhb127/gonew-test/helloserver github.com/zhb127/demo

gonew: initialized github.com/zhb127/demo in ./demo
```

but:

```bash
cd demo
./scripts/init-local-dev.sh

zsh: permission denied: ./scripts/init-local-dev.sh
```

**Reply by earthboundkid** [reply]
For my template tool, I didn't want to add a mode for setting file permissions, so I just hard coded that files ending in .sh are chmod +x. It's a tricky UI question.

**Reply by szabba** [reply]
It's not 100% clear what that should do on Windows. Both git and Git Bash had to come up with _some_ answers thought so there's prior work.

**Comment by timrulebosch** [comment]
@rsc this could be interesting, if you can keep it simple. 

Previously I've used Copier, its a templating approach which, while powerful, has the limitation that the template itself is not valid code. So, all the toolings you might use for QA and doc don't work. So, in that regard, I find gonew to be very interesting.


I would (like to) use this gonew to build projects based on existing patterns. Pull in a CLI implementation, then a service (and tests), perhaps a doc layout (which is compatible with our doc systems) and so on.

**Reply by timrulebosch** [reply]
Just tried this, pulling in a few of the exiting templates into a folder. The only observation is that gonew only works for go modules, so you can't just use it to pull in some plain content.

This is the workflow I might like to use:

```
mkdir greetings
cd greetings
go mod init example.com/greetings

gonew golang.org/x/example/helloserver example.com/greetings/internal/myserver
gonew golang.org/x/example/cli example.com/greetings/cmd/server
gonew golang.org/x/example/rest example.com/greetings/rest/server

# This bit did not work, and perhaps that is OK. You can also CURL such content.
gonew github.com/GoogleCloudPlatform/osconfig/THIRD_PARTY_LICENSES/cloud.google.com/go example.com/greetings/licenses/go
```

**Reply by earthboundkid** [reply]
Hugo uses go.mod to distribute its theme files, so there is some precedent for using go.mod for content. It is a little weird though.

**Comment by terryosayawe-cpi** [comment]
Really cool, I need to give it a try.

**Comment by trentrosenbaum** [comment]
It is really good to see gonew working over this initial interaction.  @rsc am I right in thinking that the modules being used to create a new module are meant to be fully functioning standalone and not templated themselves?  I see translating an example project into a templated syntax to be a  barrier to maintenance.  I see the original module as being a fully working standalone project and not having to be adapted for the gonew tool.

It feels that working over the go.mod and the directory name would give enough information to adapt references to internal imports that might be in the module.

It feels like in the go module landscape all modules should rename useable and not a combination of traditional modules and templated sytntax modules.

Hopefully I have  been able to explain myself.  In short I think gonew is a positive start 😄

**Comment by jimmyfrasche** [comment]
go new
1. extracts the module files locally
2. renames the module

Both of these are valuable on their own. Renaming a module obviously so. Extracting files less so, but it would be useful.† If those were separate commands then `go new` as it is would just be a helper to run those two in the correct order with the correct args.

† Consider a web framework that by convention embeds all its template files on an exported Templates variable and has a naming system that lets you override a dependency's template file. Handy on its own but you usually want to start from the original template instead of from scratch. It's already in the module cache so being able to ask go to grab just that file from the module being used is simpler than going to github and finding the file at the correct version.

requests for extensions so far generally fall into two categories
1. templating
2. refactoring

I think it's unlikely that templating will be included. Some basic package renaming would be nice but also seems unlikely given how awkward to specify this would be.

In both cases you could use some tool after you call go new to get the desired results but you may not know that you have to so there would probably be some kind of "read me first" file though you may not know to look for it. If the file name were standardized, maybe go new could cat it before exiting.

Alternately, while generally go doesn't run arbitrary code, maybe this could be the exception as it's meant to be run once by a developer. If there were a standard go tool/generate command name to run at the end it could collect any additional information and run any templating/refactoring.

**Comment by pwango** [comment]
The module web have another module file (tool includes hw.go, package name is hw, just for printing hello world )

When I used the test.go to import "web/tool", the computer still return error "package web/tool is not in std (C:\Program Files\Go\src\web\tool)" even after using go mod and go work command.

Could you tell me the way to address the error for using local module? I am just a rookie in golang, thanks.

**Comment by sysulq** [comment]
Suddenly seeing this discussion, it is very interesting that our team also developed a similar tool in the past two years and used it extensively internally. This greatly contributes to improving development efficiency, especially in the case of microservice development.

It's worth mentioning that we also supported caching and generating project structure based on module paths. However, the caching mechanism implemented in "gonew" is more elegant, we will also consider learning from it and optimizing our own implementation.

Running this command below would generate a project in `github.com/douyu/exampleapp` directory with `github.com/douyu/exampleapp` module name.

```bash
jupiter new github.com/douyu/jupiter-layout github.com/douyu/exampleapp
```
Once gonew is merged into the official Go repository, we will immediately replace and start using it. 😄 

Refer:
- https://github.com/douyu/jupiter/tree/master#quick-start
- https://github.com/douyu/jupiter/blob/master/pkg/core/cmd/cmd_new.go

**Comment by anilsenay** [comment]
I was excited to use gonew but I had a issue that I want to share it with you.
I have a repository in `bitbucket.org/company/project-name` but in project module name is `company.com/project-name` which is using in imports and go.mod file. So when I use `gonew bitbucket.org/company/project-name new-project`, it will not replace imports as `new-project` because it use module name as `bitbucket.org/company/project-name` and try to replace it with new one.

I made a quick solution for myself by cloning gonew and adding new command line flags in [this repo](https://github.com/anilsenay/gonew) but maybe you can think of this issue and find a better solution.

**Comment by jeffmccune** [comment]
To echo others who have requested input parameters, it'd be great if I as the template author could tell gonew to unmarshal yaml or json provided by the template user into a struct accessible from the templates.  The use case is to have the user provide values like a hostname or url.

**Comment by atdiar** [comment]
A quick question/feedback as I am currently implementing a project launcher CLI which is similar to gonew.
(Note: I haven't tried with gonew yet as I am waiting for the tool to get merged in)
Thinking that it might be pertinent here.

When I create a new project that will live in a workspace, seems like I  need to update the go.work file before I build the project, otherwise I get:

```text
Successfully initialized go module: github.com/atdiar/zuitest
current directory is contained in a module that is not one of the workspace modules listed in go.work. You can add the module to the workspace using:
        go work use ..
```

The first line is a custom error message, the second occurs because I am trying to build a main.go into a main.wasm

I probably have to check whether I'm in a workspace and if so, update the workspace programmatically, unless there is a better way?

cc @bcmills

**Reply by zephyrtronium** [reply]
FWIW, if you're already certain you have a Go toolchain, then detecting that you're currently in a workspace is just checking whether the output of `go env GOWORK` is non-empty. It also seems reasonable to simply always run `go work use .` and discard the error message.

Either way, generally speaking, it's preferable to invoke the go tool to manipulate the Go environment, rather than trying to reimplement its behavior.

**Reply by atdiar** [reply]
Good look! So I guess the remaining question is whether gonew should do that, I don't think that currently the code is workspace aware.

**Reply by atdiar** [reply]
Oh and still on the topic of integration with workspaces, maybe there is a way to update the workspace once a module is removed?

Like a go work tidy?

**Comment by pellared** [comment]
If the tool would not change, then I would consider renaming the tool to `gofork` as it is a neat tool for making forks of existing repositories so that it they be easily published as separate Go modules.

Without the ability to pass anything more than `dstmod` as input, I think that calling the solution as "template" is an overuse.

I think that with `gonew` you should be able to use templates that would allow you to quickly bootstrap repositories like https://github.com/golang/pkgsite (I think it could be a good test subject).

Still thanks for your great work 👍

**Reply by bcmills** [reply]
`gofork` would be misleading, I think. Instantiating a template does not preserve the commit history leading up to that template, wherease  a “fork” of a repo usually does include the history.

**Comment by pellared** [comment]
What about repositories which have multiple Go modules?

For instance it is quite common to follow https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module. You can take a look at https://github.com/goyek/goyek which has `tools` as a separate Go module. This pattern is especially important for libraries so that its users do NOT depend on the libraries' build tools. However, it could also profit for applications (e.g. CLI apps) as it should speed up the time of building the applications using `go install`. Here is an example GitHub repository template that creates two Go modules which is good for libraries https://github.com/goyek/template.

**Comment by aelse** [comment]
I want to be positive about experiments and contributions to the go ecosystem. I realise that we need to cater to developers of all levels and sometimes a problem that seems trivial to someone familiar with go may be a challenge to a developer just arriving from another language. That said, I don't see what problem gonew solves in its current state.

Looking at the [gonew source](https://cs.opensource.google/go/x/tools/+/refs/tags/v0.16.1:cmd/gonew/main.go) it's just replacing the old module name with the new name in source code. This is not a challenging problem to solve for someone who wants to copy an existing repo to a new namespace.

When i saw the announcement I had hoped for something similar to cookiecutter, which i've been using for years to help bootstrap new projects. This has been incredibly useful to my teams in crafting and reusing project templates for go and other languages.

I think the current version of gonew is effectively a conversation starter (and this discussion is the conversation!). What I hope is that gonew becomes something similar to what we're doing with cookiecutter so that we have a go native tool for bootstrapping projects and a common language for defining project templates. This would seem to be something easily built on top of text/template in conjunction with a project template specification.

Is this the direction we're likely to head?

**Comment by ghasemloo** [comment]
I think this can be useful. It can save people coding time if there are maintained standard code templates for common things like a web server, etc. similar to how some Google-internal server frameworks create the initial files.

One con to keep in mind is that making things easier to write is not always the best thing. Copying over a large amount of boilerplate code over and over without understanding and real need can make things less maintainable over time (if there is not a refactoring to some dependency that is maintain, the folks who copied the code would need to maintain the copy). 

So it might be good to think about the ecosystem and maintainability impact of having such a tool.

**Comment by mattventura** [comment]
I am looking at using this in the context of a monorepo. In this case, the templates and the destination would be in the same repo. Is there a good way to use this tool to reference a local path as the module template? Or am I over-complicating something that should just be a simple directory copy?

**Comment by iwittkau** [comment]
Is this still worth trying out?
Both example repositories from the [Google Cloud](https://github.com/GoogleCloudPlatform/go-templates) and [Service Weaver](https://github.com/ServiceWeaver/template) teams that are mentioned in the [announcement blog post](https://go.dev/blog/gonew) have been archived.

**Comment by CAFxX** [comment]
If one is allowed to dream, it would be great if such a tool allowed modules to also define migrations - not of the db variety, but rather for the code itself - so that e.g. when a new version of a template is released, the template author can also define how generated code from a previous version can be transparently updated to the new version.

When one has hundreds of modules using a template, being able to just automatically update the template to a new version would likely be a massive win.

**Comment by gopalmani** [comment]
One thing that would be extremely useful is a middle ground between the minimal module-rename behaviour and a full templating engine:
	•	basic variable replacement (module name, project name, org)
	•	optional inclusion/exclusion of files or directories
	•	a way for template authors to define simple parameters or config (similar to how users mentioned Cookiecutter or Copier)

Not expecting gonew to become a full scaffolding framework, but having these primitives would cover a lot of practical use cases without adding complexity to the template author.

**Comment by sajjanjyothi** [comment]
I have added something like **go service init service-name** in this patch set https://go-review.googlesource.com/c/go/+/733580

**Comment by pivaldi** [comment]
For anyone looking for a more template-complete alternative while gonew evolves: I've been working on [goplt](https://github.com/piprim/goplt), a cookiecutter-style scaffolding tool for Go that builds on the same module-distribution idea.

It uses go mod download to fetch templates, so you get the same versioning and proxy caching as gonew, but adds:

- Full text/template rendering for both file contents and file paths
- A template.toml manifest declaring typed variables (text, yes/no, select)
- Conditional file/directory inclusion
- Post-generate hooks (go mod tidy, git init, …) with security confirmation
- A target-dir field to auto-create a named subdirectory from a variable

`go install github.com/piprim/goplt/cmd/goplt@latest`
`goplt generate --template github.com/piprim/goplt-tmpl/cli-cobra`

Still early (v0.1.1), but functional.
Happy to hear feedback from people who've been thinking about this problem.
